hadoop-ozone/dist/src/main/compose/testlib.sh

```diff
@@ -158,13 +158,12 @@ start_docker_env(){
   export OZONE_SAFEMODE_MIN_DATANODES="${datanode_count}"
 
   docker-compose --ansi never down
-  if ! { docker-compose --ansi never up -d --scale datanode="${datanode_count}" \
-      && wait_for_safemode_exit \
-      && wait_for_om_leader ; }; then
-    [[ -n "$OUTPUT_NAME" ]] || OUTPUT_NAME="$COMPOSE_ENV_NAME"
-    stop_docker_env
-    return 1
-  fi
+
+  trap stop_docker_env EXIT HUP INT TERM
+
+  docker-compose --ansi never up -d --scale datanode="${datanode_count}"
```

Comment:
IIUC, with the change:
1. `copy_daemon_logs` and `stop_docker_env` calls are removed from `execute_robot_test()`
2. `copy_daemon_logs` is now called by `stop_docker_env()`
3. `stop_docker_env` is now called from `start_docker_env()`

Thus whoever calls `start_docker_env` **_last_** would have to call `stop_docker_env` in order to have the logs collected, right?

Reply:
Thanks @smengcl for the review.

> whoever calls `start_docker_env` **_last_** would have to call `stop_docker_env` in order to have the logs collected

It's like a `ShutdownHook` in Java.  The shell calls `stop_docker_env` when the test script exits.  `start_docker_env` just sets it up.

Reply:
ah makes sense now. That is exactly what `trap` is for. Thanks @adoroszlai .